### PR TITLE
feat(mobile): create react-native-usb library

### PR DESCRIPTION
**Pull request by Nodonisko**
Basically polyfill for WebUSB API in React Native.

TODO:

- [x] Align react-native-usb package with our structure of packages in repo
- [x] Release config for react-native-usb package because it will have to be published because it's used in transport
- [x] Proper manifest for mobile app
- [x] Add USB intent filters to suite-native app
- [x] Remove unused Kotlin files for Views
- [x] Align versions of expo-modules-core
- [x] onConnect/disconnect events works properly

Questions:
1. Should be this as optionalDependency for transport?

### CODE REVIEW

File: packages/transport/src/index.ts [hunk 1]
```diff
@@ -32,3 +32,6 @@ export { SessionsBackground } from './sessions/background';
 export { SessionsClient } from './sessions/client';
 
 export { UdpTransport } from './transports/udp';
+
+// react-native only
+export { NativeUsbTransport } from './transports/nativeUsb';
```
Comment (mroz22) [inline packages/transport/src/index.ts]: maybe it shouldn't be exported through index? 

I think it should be implemented like this

```
import { NativeUsbTransport } from '@trezor/transport/nativeUsb

TrezorConnect.init({
  transports: [new NativeUsbTransport()]
})
```

also please note that this is going to be the first usage when transport instance is passed to init, so there migh be problems. Let me know if you bump into something

File: packages/transport/package.json [hunk 2]
```diff
@@ -44,6 +44,7 @@
         "prepublish": "yarn tsx ../../scripts/prepublish.js"
     },
     "devDependencies": {
+        "@trezor/react-native-usb": "workspace:*",
```
Comment (mroz22) [inline packages/transport/package.json]: the other option would be creating separate package 

@trezor/transport-native

that would define `packages/transport/src/transports/nativeUsb.ts` using imports from `@trezor/transport`

Comment (Nodonisko) [reply packages/transport/package.json]: That makes sense, but not sure if there is any benefit because it must be imported in connect anyway to register so it will be in connect dependency tree anyway.

Only way arround is some dynamic transport injecting during connect init.

File: suite-native/state/src/extraDependencies.ts [hunk 3]
```diff
@@ -12,14 +12,29 @@ export const extraDependencies: ExtraDependencies = mergeDeepObject(extraDepende
...
     } as Partial<ExtraDependencies['selectors']>,
     thunks: {} as Partial<ExtraDependencies['thunks']>,
     actions: {
         setWalletSettingsLocalCurrency: setFiatCurrency,
     } as Partial<ExtraDependencies['actions']>,
     actionTypes: {} as Partial<ExtraDependencies['actionTypes']>,
     reducers: {} as Partial<ExtraDependencies['reducers']>,
-    utils: {} as Partial<ExtraDependencies['utils']>,
+    utils: {
+        connectInitSettings: {
```
Comment (juriczech) [inline suite-native/state/src/extraDependencies.ts:41]: These settings are already applied through `suite-common/test-utils/src/extraDependenciesMock.ts`

Could you please unify this?

File: packages/react-native-usb/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    clearHalt: createNoop('clearHalt'),
+    transferIn: (endpointNumber: number, length: number) =>
+        transferIn(device.deviceName, endpointNumber, length),
+    transferOut: (endpointNumber: number, data: BufferSource) =>
+        transferOut(device.deviceName, endpointNumber, data),
+    isochronousTransferIn: createNoop('isochronousTransferIn'),
+    isochronousTransferOut: createNoop('isochronousTransferOut'),
+    reset: createNoop('reset'),
+
+    // TODO: Implement these properties
```
Comment (juriczech) [inline packages/react-native-usb/src/index.ts]: nit: Would be great to have an issue link for this as this is a big PR and will get lost IMO

File: packages/react-native-usb/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    selectAlternateInterface: createNoop('selectAlternateInterface'),
+    controlTransferIn: createNoop('controlTransferIn'),
+    controlTransferOut: createNoop('controlTransferOut'),
+    clearHalt: createNoop('clearHalt'),
+    transferIn: (endpointNumber: number, length: number) =>
+        transferIn(device.deviceName, endpointNumber, length),
+    transferOut: (endpointNumber: number, data: BufferSource) =>
+        transferOut(device.deviceName, endpointNumber, data),
+    isochronousTransferIn: createNoop('isochronousTransferIn'),
+    isochronousTransferOut: createNoop('isochronousTransferOut'),
```
Comment (juriczech) [inline packages/react-native-usb/src/index.ts:90]: typo

File: packages/react-native-usb/src/ReactNativeUsb.types.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
+// TODO: better types than any
```
Comment (juriczech) [inline packages/react-native-usb/src/ReactNativeUsb.types.ts]: Do you plan to add types before merging?

File: packages/react-native-usb/src/ReactNativeUsb.types.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+type USBConfiguration = any;
+type USBControlTransferParameters = any;
+type USBInTransferResult = any;
+type USBOutTransferResult = any;
+type USBDirection = any;
+type USBIsochronousInTransferResult = any;
+type USBIsochronousOutTransferResult = any;
+
+export interface NativeDevice {
+    // readonly usbVersionMajor: number;
```
Comment (juriczech) [inline packages/react-native-usb/src/ReactNativeUsb.types.ts:13]: Nit: remove unused comments

File: suite-native/app/android/app/src/main/res/xml/device_filter.xml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<resources>
+    <usb-device vendor-id="21324" product-id="1" />
+    <usb-device vendor-id="4617" product-id="21440" />
+    <usb-device vendor-id="4617" product-id="21441" />
```
Comment (juriczech) [inline suite-native/app/android/app/src/main/res/xml/device_filter.xml]: Equivalent of our 3 devices?

File: packages/react-native-usb/ios/ReactNativeUsbModule.swift [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    Constants([
+      "PI": Double.pi
+    ])
+
+    // Defines event names that the module can send to JavaScript.
+    Events("onChange")
+
+    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
+    Function("hello") {
+      return "Hello world! 👋"
```
Comment (vytick) [inline packages/react-native-usb/ios/ReactNativeUsbModule.swift]: should this file (or whole ios support) be here at all? And if so, maybe without this method and statement,  that it is empty for a reson?

File: packages/react-native-usb/src/ReactNativeUsb.types.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    // readonly configurations: USBConfiguration[];
+    readonly opened: boolean;
+    // extra properties
+    readonly deviceName: string;
+}
+
+export interface WebUSBDevice {
+    readonly usbVersionMajor: number;
+    readonly usbVersionMinor: number;
+    readonly usbVersionSubminor: number;
```
Comment (vytick) [inline packages/react-native-usb/src/ReactNativeUsb.types.ts:38]: subminor -> patch ?

File: packages/react-native-usb/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+### Add the package to your npm dependencies
+
+```
+npm install react-native-usb
+```
+
+### Configure for iOS
+
+Run `npx pod-install` after installing the npm package.
```
Comment (vytick) [inline packages/react-native-usb/README.md:22]: I would add here, that it is not implemented.

File: packages/transport/src/index.ts (hunk 1 above)
Comment (Nodonisko) [reply packages/transport/src/index.ts]: What's the benefit? Generaly is not recommended to access anything in package that is not exported through index (same as private properties)

File: suite-native/state/src/extraDependencies.ts (hunk 3 above)
Comment (Nodonisko) [reply suite-native/state/src/extraDependencies.ts:41]: Mock is just mock. It's used for tests both in desktop and mobile, no reason to unify that

File: packages/react-native-usb/src/index.ts (hunk 4 above)
Comment (Nodonisko) [reply packages/react-native-usb/src/index.ts]: Good idea, but I probably wouldn't do that for this because it's super low priority because we don't using it anywhere anyway.

File: packages/react-native-usb/src/ReactNativeUsb.types.ts (hunk 6 above)
Comment (Nodonisko) [reply packages/react-native-usb/src/ReactNativeUsb.types.ts]: Probably not, we don't use these properties anywhere but I want them to be here if someone else would want to add/use them in future. This library could be used even with 3rd parties for different devices than Trezors.

File: suite-native/app/android/app/src/main/res/xml/device_filter.xml (hunk 8 above)
Comment (Nodonisko) [reply suite-native/app/android/app/src/main/res/xml/device_filter.xml]: Good catch, TS3 is missing here, first two are T1 I think. I will verify this once more.

File: packages/react-native-usb/ios/ReactNativeUsbModule.swift (hunk 9 above)
Comment (Nodonisko) [reply packages/react-native-usb/ios/ReactNativeUsbModule.swift]: Good catch, I will add some comment.

File: packages/react-native-usb/src/ReactNativeUsb.types.ts (hunk 10 above)
Comment (Nodonisko) [reply packages/react-native-usb/src/ReactNativeUsb.types.ts:38]: It's from definition of WebUSB

Comment (mroz22) [reply packages/react-native-usb/src/ReactNativeUsb.types.ts:38]: my guess is dan was following webusb device shape from browser?

File: packages/react-native-usb/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
+/* eslint-disable no-console */
+import { NativeModulesProxy, EventEmitter, Subscription } from 'expo-modules-core';
+
+import { ReactNativeUsbModule } from './ReactNativeUsbModule';
+import { NativeDevice, OnConnectEvent, WebUSBDevice } from './ReactNativeUsb.types';
+
+const DEBUG_LOGS = true;
```
Comment (PeKne) [inline packages/react-native-usb/src/index.ts:7]: Should not this be `false` by default?

File: packages/transport/src/transports/nativeUsb.native.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+// notes:
+// to make it work on Linux I needed to run `sudo chmod -R 777 /dev/bus/usb/` which is obviously not
+// the way to go.
+
+export class NativeUsbTransport extends AbstractUsbTransport {
+    public name = 'NativeUsbTransport' as const;
+
+    constructor({ messages, logger }: ConstructorParameters<typeof AbstractTransport>[0]) {
+        console.log('NativeUsbTransport constructor');
```
Comment (PeKne) [inline packages/transport/src/transports/nativeUsb.native.ts]: forgotten console.log

File: packages/react-native-usb/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+const transferIn = async (deviceName: string, endpointNumber: number, length: number) => {
+    const perf = performance.now();
+    const data = await ReactNativeUsbModule.transferIn(deviceName, endpointNumber, length)
+        .catch((error: any) => {
+            debugLog('JS: USB read error: ', error);
+            throw error;
+        })
+        // eslint-disable-next-line arrow-body-style
+        .then((result: number[]) => {
+            // debugLog('JS: Native USB read result:', JSON.stringify(result));
```
Comment (PeKne) [inline packages/react-native-usb/src/index.ts]: uncomment or delete this

File: packages/react-native-usb/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    usbVersionMinor: 0,
+    usbVersionSubminor: 0,
+    deviceVersionMajor: 1,
+    deviceVersionMinor: 0,
+    deviceVersionSubminor: 0,
+    configurations: [],
+});
+
+export function onDeviceConnected(listener: (event: OnConnectEvent) => void): Subscription {
+    return emitter.addListener<NativeDevice>('onDeviceConnect', event => {
```
Comment (PeKne) [inline packages/react-native-usb/src/index.ts:104]: Are these listeners cleared anywhere?

File: packages/react-native-usb/src/index.ts (hunk 12 above)
Comment (Nodonisko) [reply packages/react-native-usb/src/index.ts:7]: Later yes, but I expect this to be quite useful in first few days/weeks after merge so I plan to change it to false later

File: packages/react-native-usb/src/index.ts (hunk 15 above)
Comment (Nodonisko) [reply packages/react-native-usb/src/index.ts:104]: No, this is purely proxied to "client"

Comment (Nodonisko) [reply packages/react-native-usb/src/index.ts:104]: There should be probably something like removeListeners but since tranport/connect doesnt use them anyway I did not implement it

File: packages/react-native-usb/src/index.ts (hunk 4 above)
Comment (juriczech) [reply packages/react-native-usb/src/index.ts]: I would consider removing the comment then? There is a lot of comments and todos so would be worth cleaning up a little bit if these things are low prio, let's just clean it up and push it as it is

File: packages/react-native-usb/src/index.ts (hunk 5 above)
Comment (Nodonisko) [reply packages/react-native-usb/src/index.ts:90]: I am blind, where is typo? :D

File: packages/react-native-usb/src/index.ts (hunk 4 above)
Comment (Nodonisko) [reply packages/react-native-usb/src/index.ts]: I added more extensive messages that explains why they are there and why it's low prio

File: packages/react-native-usb/README.md (hunk 11 above)
Comment (Nodonisko) [reply packages/react-native-usb/README.md:22]: done

**Review by PeKne — APPROVED** [review]

**Review by juriczech — APPROVED** [review]

File: packages/react-native-usb/ios/ReactNativeUsbModule.swift [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  // Each module class must implement the definition function. The definition consists of components
+  // that describes the module's functionality and behavior.
+  // See https://docs.expo.dev/modules/module-api for more details about available components.
+  public func definition() -> ModuleDefinition {
+    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
+    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
+    // The module will be accessible from `requireNativeModule('ReactNativeUsb')` in JavaScript.
+    Name("ReactNativeUsb")
+
+    // Sets constant properties on the module. Can take a dictionary or a closure that returns a dictionary.
```
Comment (vytick) [inline packages/react-native-usb/ios/ReactNativeUsbModule.swift:15]: Maybe it would be nice to remove all commented code for the information clarity

**Review by vytick — APPROVED** [review]

File: packages/react-native-usb/ios/ReactNativeUsbModule.swift (hunk 16 above)
Comment (Nodonisko) [reply packages/react-native-usb/ios/ReactNativeUsbModule.swift:15]: I want to keep it, will be easier to pick up with examles if someone decide to do iOS

File: package.json [hunk 17]
```diff
@@ -88,7 +90,8 @@
         "css-loader": "^6.8.1",
         "html-webpack-plugin": "^5.5.3",
         "@types/node": "18.17.15",
-        "@sentry/types": "7.56.0"
+        "@sentry/types": "7.56.0",
+        "@types/react": "18.0.32"
```
Comment (mroz22) [inline package.json:94]: changes in this file. are they related? needed?

File: suite-native/state/src/extraDependencies.ts [hunk 18]
```diff
@@ -1,25 +1,54 @@
...
 import { selectDevices } from '@suite-common/wallet-core';
 import { selectFiatCurrencyCode, setFiatCurrency } from '@suite-native/module-settings';
 import { PROTO } from '@trezor/connect';
 import { mergeDeepObject } from '@trezor/utils';
+import { NativeUsbTransport } from '@trezor/transport-native';
+
+const protobugMessages = require('@trezor/protobuf/messages.json');
+
+const transports = Platform.select({
+    ios: ['UdpTransport'],
```
Comment (mroz22) [inline suite-native/state/src/extraDependencies.ts:15]: udp transport is basically emulator without bridge.

File: suite-native/state/src/extraDependencies.ts [hunk 19]
```diff
@@ -1,25 +1,54 @@
...
+    android: [
+        new NativeUsbTransport({
+            messages: protobugMessages,
```
Comment (mroz22) [inline suite-native/state/src/extraDependencies.ts]: nic typo-bug `protobug` :D. I would prefer assigning `'@trezor/protobuf/messages.json'` as default inside `AbstractTransport` in transport package. But as long this works too, no problem with it as it is. 

tldr - we might need in the future to support passing custom messages to transport (because of some shitcoin-plugins-modularization), but this would be happening on connect level anyway.

**Review by mroz22 — APPROVED** [review]
Nice, minimum changes on transport/connect level. If it works,I am really happy

File: suite-native/state/src/extraDependencies.ts (hunk 18 above)
Comment (Nodonisko) [reply suite-native/state/src/extraDependencies.ts:15]: I know we want that on iOS for testing

File: package.json (hunk 17 above)
Comment (Nodonisko) [reply package.json:94]: Had some issues with types for React, not sure why now but this shoudn't hurt

File: suite-native/state/src/extraDependencies.ts (hunk 19 above)
Comment (Nodonisko) [reply suite-native/state/src/extraDependencies.ts]: I will keep it here now since I don't want to complicate things in this PR.

typo fixed :D
